### PR TITLE
3053 recursive objects

### PR DESCRIPTION
**Pull request by easherma**
# Description

I tried a couple approaches, this one leverages the 'stream_data' object which ultimately returns a JSON-like dict from the Streamfields. 

It works for our current use cases, but ought to be tightened up a bit to be a bit more foolproof. 

Fixes # https://github.com/cityofaustin/techstack/issues/3053

# Testing Notes

See if links work in janis, or use graphiql

* [Deployed Link]()

# Checklist:
- [ ] Request reviewers
- [ ] Slack-out message for review
- [ ] Link this PR in the issue
- [ ] Moved card to *review* in zenhub

### CODE REVIEW

File: joplin/api/schema.py [hunk 1]
```diff
@@ -6,19 +6,104 @@
...
+                    block = item.block
+                    try:
+                        if isinstance(block, RichTextBlock):
+                            rt_names.append(block.name)
+                        elif block.child_blocks:
+                            for child_block in block.all_blocks():
+                                if isinstance(child_block, RichTextBlock):
+                                    rt_names.append(child_block.name)
+                    except AttributeError as e:
+                        pass
```
Comment (easherma) [inline joplin/api/schema.py]: This should be refactored into its own function

File: joplin/api/schema.py [hunk 2]
```diff
@@ -6,19 +6,104 @@
...
+            2) use each of the above names as keys to sort through the stream_data
+                (which is a python dict represention of the StreamField, alot like JSON)
+                use the nested_lookup to recursivley return a list of all values that match the key
+                run a nested_alter using expand_db_html as a callback to edit those values in stream_data
+                (otherwise the values tend to get overwritten)
+
+        """
+        print(type(StreamValue))
+        try:
+            data = StreamValue.stream_data
```
Comment (easherma) [inline joplin/api/schema.py]: dosen't get used til line 80...sorry

File: joplin/api/schema.py [hunk 3]
```diff
@@ -6,19 +6,104 @@
...
+
+                    for value in values:
+                        if not isinstance(value, str):
+                            for elem in value:
+                                for key in rt_names:
+                                    print('alter elem', key)
+                                    altered_value = nested_alter(elem, key, expand_db_html)
+                # fallback to 'value'
+                for elem in data:
+                    if elem['type'] == 'basic_step':
```
Comment (easherma) [inline joplin/api/schema.py]: Alot of the initial attempt at this was to have a function that dynamically got all the cases needed from our code to alter, so that we wouldn't have to revisit every time we do something fancy with StreamFields. 
However it was pretty hard to get this working due to the complex data structures, so I ended up hard-coding in some use cases to make it functional. 
So now its kind of a mix of generally reusable stuff and hard-coded stuff, which is...not great.

File: Pipfile [hunk 4]
```diff
@@ -13,7 +13,7 @@ django-filter = "==2.1.0"
 django-modeltranslation = "==0.13.1"
 wagtail-modeltranslation = "==0.10.2"
 dj-database-url = "==0.5.0"
-graphene-django = "==2.2.0"
+graphene-django = "==2.6.0"
```
Comment (briaguya) [inline Pipfile]: note to self mainly, what's new in 2.6? Are there things in the docs that make some API stuff easier/cleaner?

File: README.md [hunk 5]
```diff
@@ -347,6 +347,7 @@ HEROKU_JANIS_APP_NAME=janis-staging ./scripts/serve-local.sh
 ```
 
 #### Syncing prod data to staging
-1. Create a new seeding data backup sourced from prod, using `migration_test.sh`
-2. Drop the staging database (go to heroku and delete the database on staging) 
+
+1. Create a new seeding data backup sourced from prod, using `LOAD_PROD_DATA=on migration_test.sh`
```
Comment (briaguya) [inline README.md]: note: when created from prod data we see that it was sourced from prod in the [seeding_datadump_metadata.txt](https://github.com/cityofaustin/joplin/blob/f847c0b06947c4d4a04160573ff10a429755e9f7/joplin/db/system-generated/seeding_datadump_metadata.txt#L7)

File: joplin/api/schema.py [hunk 6]
```diff
@@ -6,19 +6,116 @@
...
+from wagtail.core.rich_text import expand_db_html
 from base.models import TranslatedImage, ThreeOneOne, ServicePage, ServicePageContact, ServicePageTopic, ServicePageRelatedDepartments, InformationPageRelatedDepartments, ProcessPage, ProcessPageStep, ProcessPageContact, ProcessPageTopic, InformationPage, InformationPageContact, InformationPageTopic, DepartmentPage, DepartmentPageContact, DepartmentPageDirector, Theme, TopicCollectionPage, TopicPage, Contact, Location, ContactDayAndDuration, Department, DepartmentContact, TopicPageTopicCollection, OfficialDocumentPage, OfficialDocumentPageRelatedDepartments, OfficialDocumentPageTopic, OfficialDocumentPageOfficialDocument, GuidePage, GuidePageTopic, GuidePageRelatedDepartments, GuidePageContact, JanisBasePage, PhoneNumber, DepartmentPageTopPage, DepartmentPageRelatedPage
+import collections
+import traceback
+from nested_lookup import nested_lookup, nested_alter, nested_update, get_occurrence_of_key, get_occurrence_of_value
+
+
+class RichTextFieldType(Scalar):
+    """Serialises RichText content into fully baked HTML
+    see https://github.com/wagtail/wagtail/issues/2695#issuecomment-373002412 """
```
Comment (briaguya) [inline joplin/api/schema.py]: went in a loop because this is copied from https://github.com/wagtail/wagtail/issues/2695#issuecomment-464462575

File: joplin/api/schema.py [hunk 7]
```diff
@@ -6,19 +6,116 @@
...
+                loop through all the values and try running expand_db_html on them to return an href
+                still need to find a way to actually attach the janis_url to the selection tho..
+    """
+
     @staticmethod
-    def serialize(dt):
-        return [{'type': item.block_type, 'value': item.block.get_api_representation(item.value), 'id': item.id} for item in dt]
+    def serialize(StreamValue):
+        """
+        This is a rats nest of loops and conditionals, but the general idea is two functions:
```
Comment (briaguya) [inline joplin/api/schema.py]: It's quite hard to parse right now, but could be pretty easily cleaned up by moving a couple chunks into named methods.

File: joplin/api/schema.py [hunk 8]
```diff
@@ -6,19 +6,116 @@
...
+                    """
+                    if the value isnt a string, there is another nested list
+                    we need to alter each element of that list on its own to
+                    avoid accidentially overriding values
+                    """
+                    if not isinstance(value, str):
+                        for elem in value:
+                            for key in rich_text_names:
+                                print('alter elem', key)
+                                altered_value = nested_alter(elem, key, expand_db_html)
```
Comment (briaguya) [inline joplin/api/schema.py]: moving this into a named method would be huge

File: joplin/api/schema.py [hunk 9]
```diff
@@ -6,19 +6,116 @@
...
+            print(traceback.format_exc())
+            pass
+
+            # todo: make things less nested
+        for block_key in rich_text_names:
+            if not isinstance(StreamValue, str):
+                data = StreamValue.stream_data
+            else:
+                data = ''
+            # sometimes data is just a string, and there isno stream_data
```
Comment (briaguya) [inline joplin/api/schema.py]: This reads better to me above block getting data from StreamValue, as it seems to apply to both that and the check below

File: joplin/api/schema.py (hunk 8 above)
Comment (briaguya) [inline joplin/api/schema.py]: or just the inner part of the loop, up to you

File: joplin/base/wagtail_hooks.py [hunk 10]
```diff
@@ -237,3 +241,33 @@ def show_live_pages_only(pages, request):
...
+    def get_model():
+        return Page
+
+    @classmethod
+    def get_instance(cls, attrs):
+        return super().get_instance(attrs).specific
+
+    @classmethod
+    def expand_db_attributes(cls, attrs):
+        print("expanding!")
```
Comment (briaguya) [inline joplin/base/wagtail_hooks.py]: just a note to clean this up

File: scripts/remove_logs_from_json_stream.py [hunk 11]
```diff
@@ -11,11 +13,13 @@
 '''
 
 start_of_json_pattern = re.compile("^\[")
-in_json=False
+in_json = False
 
 for line in fileinput.input():
+    import pdb
+    pdb.set_trace()
```
Comment (briaguya) [inline scripts/remove_logs_from_json_stream.py]: note to cleanup

File: Pipfile (hunk 4 above)
Comment (easherma) [reply Pipfile]: nothing major that I noticed, but figured i might as well update the dependency

File: joplin/api/schema.py (hunk 6 above)
Comment (easherma) [reply joplin/api/schema.py]: I left it in, I still think its a relevant issue to note. I could change it so its just the last comment on that issue though?

File: joplin/api/schema.py (hunk 8 above)
Comment (easherma) [reply joplin/api/schema.py]: I'm not sure what you mean

File: joplin/api/schema.py (hunk 6 above)
Comment (briaguya) [reply joplin/api/schema.py]: that would be perfect.

File: joplin/api/schema.py (hunk 8 above)
Comment (briaguya) [reply joplin/api/schema.py]:
```
for value in values:
   do_a_function_thing(value)

File: joplin/base/models/janis_page.py [hunk 12]
```diff
@@ -113,7 +114,7 @@ def janis_url(self):
             )
 
             # add hardcoded language path to base url
-            base_url = os.environ["JANIS_URL"] + '/en'
+            base_url = f"{self.janis_url_base('publish_janis_branch')}/en"
```
Comment (briaguya) [inline joplin/base/models/janis_page.py]: yay!

File: joplin/base/models/janis_page.py [hunk 13]
```diff
@@ -175,20 +167,31 @@ def janis_preview_url_end(self, revision=None):
...
+        returns a valid url of the base URL in janis:
+            Use hardcoded JANIS_URL for staging and prod
+            Otherwise, use configurable branch setting
+        """
         if settings.ISSTAGING or settings.ISPRODUCTION:
             return os.getenv("JANIS_URL")
         else:
-            return JanisBranchSettings.objects.first().branch_preview_url_base()
+            branch_settings = JanisBranchSettings.objects.first()
+            return branch_settings.url_base(branch)
```
Comment (briaguya) [inline joplin/base/models/janis_page.py]: it seems like this line evaluates to:
`return branch_settings.url_base('preview_janis_branch')`

seems a little magic stringy, or maybe the var name branch is throwing me for a loop, not a dealbreaker but something to ponder on

Comment (easherma) [reply joplin/base/models/janis_page.py]: I wrote it this way so that you can use either the preview or publish branch, like on line 117 above

File: joplin/base/models/site_settings.py [hunk 14]
```diff
@@ -41,9 +41,18 @@ class JanisBranchSettings(BaseSetting):
...
             return self.preview_janis_url
         else:
-            # Netlify site names are limited to 63 characters
-            return f"https://{('janis-' + self.preview_janis_branch)[:63]}.netlify.com"
+            try:
+                # Netlify site names are limited to 63 characters
+                return f"https://{('janis-' + getattr(self, janis_branch))[:63]}.netlify.com"
+            except AttributeError as e:
+                # TODO: better handiling, I think this defaults back to staging/prod as is
+                print("something missing", e)
```
Comment (Niicck) [inline joplin/base/models/site_settings.py]: In the future, `except` clauses should handle errors or not exist at all. Showing a message as vague as "something missing" or "something wrong with flags maybe" is unhelpful or worse, passing breaking errors that should be addressed. 

It's a very good idea to handle potential errors here. But printing a statement and passing is not error handling. If error handling is a feature requirement that you don't want to tackle in this same PR, make a github issue to address it.

File: joplin/base/models/site_settings.py [hunk 15]
```diff
@@ -41,9 +41,18 @@ class JanisBranchSettings(BaseSetting):
     ]
 
     # Convert branch name to a netlify URL
-    def branch_preview_url_base(self):
+    def url_base(self, janis_branch):
```
Comment (Niicck) [inline joplin/base/models/site_settings.py]: `janis_branch` is a confusing name for me. It took me a minute to figure out that its a flag to get either the "publish_janis_branch" or the "preview_janis_branch"/"preview_janis_url". It's not the actual name of the janis_branch, which confused me as I read other parts of the code. I like that you've made url_base() work for both previews and publishes, but we need a better name for the janis_branch param.

File: joplin/base/models/janis_page.py [hunk 16]
```diff
@@ -175,20 +167,34 @@ def janis_preview_url_end(self, revision=None):
...
+        """
         if settings.ISSTAGING or settings.ISPRODUCTION:
             return os.getenv("JANIS_URL")
         else:
-            return JanisBranchSettings.objects.first().branch_preview_url_base()
+            branch_settings = JanisBranchSettings.objects.first()
+            return branch_settings.url_base(janis_branch)
+
+    # alias for url base function
+    janis_preview_url_start = janis_url_base
```
Comment (Niicck) [inline joplin/base/models/janis_page.py]: The `janis_preview_url_start` function can be gotten rid of. Just use janis_url_base now that we've got it.

File: joplin/base/models/janis_page.py (hunk 12 above)
Comment (Niicck) [inline joplin/base/models/janis_page.py]: yay!

File: joplin/base/models/site_settings.py (hunk 15 above)
Comment (briaguya) [reply joplin/base/models/site_settings.py]: it was just "branch" and was v confusing. Any improvements to code understandability here would be awesome

File: joplin/base/models/site_settings.py (hunk 14 above)
Comment (briaguya) [reply joplin/base/models/site_settings.py]: we're still printing the error here, not sure if print is the best way to handle logging though, probably a bigger thing to think about

File: joplin/base/models/site_settings.py (hunk 15 above)
Comment (easherma) [reply joplin/base/models/site_settings.py]: for part of https://github.com/cityofaustin/techstack/issues/3185 I was going to see what it would take to convert these statments to use djangos logging. That way they can be controlled based on env settings and various verbosity stuff

File: joplin/base/models/site_settings.py (hunk 14 above)
Comment (easherma) [reply joplin/base/models/site_settings.py]: for part of https://github.com/cityofaustin/techstack/issues/3185 I was going to see what it would take to convert these statments to use djangos logging. That way they can be controlled based on env settings and various verbosity stuff

File: joplin/base/models/site_settings.py (hunk 15 above)
Comment (easherma) [reply joplin/base/models/site_settings.py]: "janis_branch_type" ?
